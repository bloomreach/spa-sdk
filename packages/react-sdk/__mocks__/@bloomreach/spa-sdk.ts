/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
