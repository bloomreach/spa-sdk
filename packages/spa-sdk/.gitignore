/coverage
/docs
/lib
/node_modules
.DS_store