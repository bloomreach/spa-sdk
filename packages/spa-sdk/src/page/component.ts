/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MetaCollectionModel, Meta } from './meta';

/**
 * Generic component type.
 */
export const TYPE_COMPONENT = 'COMPONENT';

/**
 * Parameters of a component.
 */
export interface ComponentParameters {
  [name: string]: string | undefined;
}

/**
 * Meta-data of a component.
 */
export interface ComponentMeta extends MetaCollectionModel {
  params?: ComponentParameters;
}

interface Models {
  [model: string]: any;
}

interface Links {
  componentRendering?: { href: string };
}

/**
 * Model of a component.
 */
export interface ComponentModel {
  _meta?: ComponentMeta;
  _links?: Links;
  id: string;
  models?: Models;
  name?: string;
  type: typeof TYPE_COMPONENT | string;
  components?: ComponentModel[];
}

/**
 * A component in the current page.
 */
export interface Component {
  /**
   * @return The component meta-data collection.
   */
  getMeta(): Meta[];

  /**
   * @return The map of models.
   */
  getModels(): Models;

  /**
   * @return The link to the partial component model.
   */
  getModelUrl(): string | undefined;

  /**
   * @return The name of this component.
   */
  getName(): string;

  /**
   * @return the parameters of this component.
   */
  getParameters(): ComponentParameters;

  /**
   * Looks up for a nested component.
   * @param componentNames A lookup path.
   */
  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;

  /**
   * Looks up for a nested component by its id.
   * @param id A component id.
   */
  getComponentById<T extends Component>(id: string): T | undefined;
}

export class Component implements Component {
  constructor(
    protected model: ComponentModel,
    protected children: Component[] = [],
    protected meta: Meta[] = [],
  ) {}

  protected getId() {
    return this.model.id;
  }

  getMeta() {
    return this.meta;
  }

  getModels() {
    return this.model.models || {};
  }

  getModelUrl() {
    return this.model._links && this.model._links.componentRendering && this.model._links.componentRendering.href;
  }

  getName() {
    return this.model.name || '';
  }

  getParameters() {
    return this.model._meta && this.model._meta.params || {};
  }

  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;
  getComponent(...componentNames: string[]) {
    // tslint:disable-next-line:no-this-assignment
    let component: Component | undefined = this;

    while (componentNames.length && component) {
      const name = componentNames.shift()!;
      component = component.children.find(component => component.getName() === name);
    }

    return component;
  }

  getComponentById(id: string) {
    const queue = [this as Component];

    while (queue.length) {
      const component = queue.shift()!;

      if (component.model.id === id) {
        return component;
      }

      queue.push(...component.children);
    }
  }
}
