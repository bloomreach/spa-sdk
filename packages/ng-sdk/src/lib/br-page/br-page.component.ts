/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AfterContentChecked,
  ChangeDetectorRef,
  ChangeDetectionStrategy,
  Component,
  Input,
  OnChanges,
  OnDestroy,
  SimpleChanges,
  TemplateRef,
  Type,
  ViewChild,
} from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { destroy, initialize, isPage, Configuration, Page, PageModel } from '@bloomreach/spa-sdk';
import { from } from 'rxjs';
import { map } from 'rxjs/operators';
import { BrComponentContext } from '../br-component.directive';
import { BrProps } from '../br-props.model';

interface BrNodeContext extends BrComponentContext {
  template?: TemplateRef<BrComponentContext>;
}

/**
 * The brXM page.
 */
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'br-page',
  templateUrl: './br-page.component.html',
})
export class BrPageComponent implements AfterContentChecked, OnChanges, OnDestroy {
  /**
   * The configuration of the SPA SDK.
   * @see https://www.npmjs.com/package/@bloomreach/spa-sdk#configuration
   */
  @Input() configuration!: Omit<Configuration, 'httpClient'>;

  /**
   * The brXM and Angular components mapping.
   */
  @Input() mapping: Record<string, Type<BrProps>> = {};

  /**
   * The pre-initialized page instance or prefetched page model.
   * Mostly this property should be used to transfer state from the server-side to the client-side.
   */
  @Input() page?: Page | PageModel;

  @ViewChild('brNode') node!: TemplateRef<BrNodeContext>;

  private instance?: Page = undefined;
  private isSynced = false;

  constructor(private changeDetectorRef: ChangeDetectorRef, private httpClient: HttpClient) {
    this.request = this.request.bind(this);
  }

  get context(): BrNodeContext | undefined {
    const component = this.state?.getComponent();

    if (!component) {
      return;
    }

    return {
      component,
      $implicit: component,
      // tslint:disable-next-line: no-non-null-assertion
      page: this.state!,
    };
  }

  ngAfterContentChecked(): void {
    if (this.isSynced) {
      return;
    }

    this.state?.sync();
    this.isSynced = true;
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes.configuration?.currentValue !== changes.configuration?.previousValue
      || changes.page?.currentValue !== changes.page?.previousValue
    ) {
      this.ngOnDestroy();
      this.initialize(changes.page?.currentValue === changes.page?.previousValue);
    }
  }

  ngOnDestroy() {
    if (!this.instance) {
      return;
    }

    destroy(this.instance);
    delete this.instance;
  }

  get state(): Page | undefined {
    return this.instance;
  }

  set state(value: Page | undefined) {
    this.instance = value;
    this.isSynced = false;
  }

  private initialize(force: boolean): void {
    const page = force ? undefined : this.page;

    if (isPage(page)) {
      this.state = page;

      return;
    }

    from(initialize({ httpClient: this.request, ...this.configuration } as Configuration, page))
      .subscribe(state => {
        this.state = state;
        this.changeDetectorRef.detectChanges();
      });
  }

  private request(...[{ data: body, headers, method, url }]: Parameters<Configuration['httpClient']>) {
    return this.httpClient.request<PageModel>(
      method,
      url,
      {
        body,
        headers: headers as Record<string, string | string[]>,
        responseType: 'json'
      },
    )
    .pipe(map(data => ({data})))
    .toPromise();
  }
}
