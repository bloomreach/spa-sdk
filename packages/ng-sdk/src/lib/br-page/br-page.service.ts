/*
 * Copyright 2023 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable, TemplateRef, Type } from '@angular/core';
import { Page } from '@bloomreach/spa-sdk';
import { BehaviorSubject } from 'rxjs';
import type { BrComponentContext } from '../br-component.directive';
import { BrProps } from '../br-props.model';

export interface BrNodeContext extends BrComponentContext {
  template?: TemplateRef<BrComponentContext>;
}

@Injectable({
  providedIn: 'root',
})
export class BrPageService {
  private _mapping: Record<keyof any, Type<BrProps>> = {};
  private _node!: TemplateRef<BrNodeContext>;
  private readonly state$: BehaviorSubject<Page | undefined>;

  constructor() {
    this.state$ = new BehaviorSubject<Page | undefined>(undefined);
  }

  get mapping(): Record<keyof any, Type<BrProps>> {
    return this._mapping;
  }

  set mapping(mapping: Record<keyof any, Type<BrProps>>) {
    this._mapping = mapping;
  }

  get node(): TemplateRef<BrNodeContext> {
    return this._node;
  }

  set node(node: TemplateRef<BrNodeContext>) {
    this._node = node;
  }

  get state(): BehaviorSubject<Page | undefined> {
    return this.state$;
  }
}
