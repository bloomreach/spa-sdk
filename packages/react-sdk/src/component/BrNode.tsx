/*
 * Copyright 2019-2020 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';
import { BrComponentContext } from './BrComponentContext';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrNodeComponent } from './BrNodeComponent';

interface BrNodeProps {
  component?: Component;
}

export class BrNode extends React.Component<BrNodeProps> {
  static contextType = BrPageContext;

  context: React.ContextType<typeof BrPageContext>;

  private renderNode(): JSX.Element | React.ReactNode {
    const { children, component } = this.props;

    if (React.Children.count(children)) {
      return <BrMeta meta={component?.getMeta()}>{children}</BrMeta>;
    }

    // eslint-disable-next-line react/no-array-index-key
    const childrenList = component?.getChildren().map((child, index) => <BrNode key={index} component={child} />);

    if (isContainer(component)) {
      return (
        <BrNodeContainer component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainer>
      );
    }

    if (isContainerItem(component)) {
      return (
        <BrNodeContainerItem component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainerItem>
      );
    }

    return (
      <BrNodeComponent component={component} page={this.context!}>
        {childrenList}
      </BrNodeComponent>
    );
  }

  render(): JSX.Element {
    const { component } = this.props;

    return <BrComponentContext.Provider value={component}>{this.renderNode()}</BrComponentContext.Provider>;
  }
}
