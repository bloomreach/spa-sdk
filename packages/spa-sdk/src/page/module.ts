/*
 * Copyright 2019-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { ContainerModule } from 'inversify';

import { CmsEventBus, CmsEventBusService, CmsEventBusServiceProvider } from '../cms';
import { UrlBuilder, UrlBuilderService } from '../url';

import { ButtonFactory } from './button-factory';
import { createManageContentButton, TYPE_MANAGE_CONTENT_BUTTON } from './button-manage-content';
import {
  ComponentChildrenToken,
  ComponentImpl,
  ComponentModelToken,
  TYPE_COMPONENT,
  TYPE_COMPONENT_CONTAINER,
  TYPE_COMPONENT_CONTAINER_ITEM,
} from './component';
import { ComponentFactory } from './component-factory';
import { ContainerImpl } from './container';
import { ContainerItemImpl } from './container-item';
import { ContentFactory } from './content-factory';
import { DocumentImpl, DocumentModelToken, TYPE_DOCUMENT } from './document';
import { PageEventBusService } from './page-events';
import { ImageFactory, ImageImpl, ImageModel, ImageModelToken } from './image';
import { ImageSetImpl, ImageSetModelToken, TYPE_IMAGE_SET } from './image-set';
import { TYPE_LINK_INTERNAL } from './link';
import { LinkFactory } from './link-factory';
import {
  DomParserServiceProvider,
  LinkRewriterImpl,
  LinkRewriterService,
  XmlSerializerServiceProvider,
} from './link-rewriter';
import { Menu, MenuImpl, MenuModelToken, TYPE_MANAGE_MENU_BUTTON, TYPE_MENU } from './menu';
import { MenuItemFactory, MenuItemImpl, MenuItemModel, MenuItemModelToken } from './menu-item';
import { TYPE_META_COMMENT } from './meta';
import { MetaCollectionImpl, MetaCollectionModel, MetaCollectionModelToken } from './meta-collection';
import { MetaCollectionFactory } from './meta-collection-factory';
import { MetaCommentImpl } from './meta-comment';
import { MetaFactory } from './meta-factory';
import { PageImpl, PageModel, PageModelToken } from './page';
import { PageFactory } from './page-factory';
import { PaginationImpl, PaginationModelToken, TYPE_PAGINATION } from './pagination';
import {
  PaginationItemFactory,
  PaginationItemImpl,
  PaginationItemModel,
  PaginationItemModelToken,
} from './pagination-item';

export function PageModule(): ContainerModule {
  return new ContainerModule((bind) => {
    bind(PageEventBusService)
      .toDynamicValue(() => new Typed())
      .inSingletonScope()
      .when(() => typeof window !== 'undefined');

    /*
     Its necessary to use a async provider here because we can only get the CmsEventBus once the module is installed,
     if the page is in preview mode
    */
    bind(CmsEventBusServiceProvider).toProvider<CmsEventBus | undefined>(
      (context) => () =>
        new Promise<CmsEventBus | undefined>((resolve) => {
          try {
            if (context.container.isBound(CmsEventBusService)) {
              const cmsEventBus = context.container.get<CmsEventBus>(CmsEventBusService);
              resolve(cmsEventBus);
            }
          } catch (e) {
            resolve(undefined);
          }
        }),
    );

    bind(LinkRewriterService).to(LinkRewriterImpl).inSingletonScope();
    bind(DomParserServiceProvider).toProvider(() => async () => {
      const { DOMParser } = await import('@xmldom/xmldom');
      return new DOMParser();
    });
    bind(XmlSerializerServiceProvider).toProvider(() => async () => {
      const { XMLSerializer } = await import('@xmldom/xmldom');
      return new XMLSerializer();
    });

    bind(ButtonFactory)
      .toSelf()
      .inSingletonScope()
      .onActivation((context, factory) =>
        factory
          .register(TYPE_MANAGE_CONTENT_BUTTON, createManageContentButton)
          .register(TYPE_MANAGE_MENU_BUTTON, (menu: Menu) => menu.getMeta()),
      );

    bind(LinkFactory)
      .toSelf()
      .inSingletonScope()
      .onActivation(({ container }, factory) => {
        const url = container.get<UrlBuilder>(UrlBuilderService);

        return factory.register(TYPE_LINK_INTERNAL, url.getSpaUrl.bind(url));
      });

    bind(MetaCollectionFactory).toFactory(({ container }) => (model: MetaCollectionModel) => {
      const scope = container.createChild();
      scope.bind(MetaCollectionImpl).toSelf();
      scope.bind(MetaCollectionModelToken).toConstantValue(model);

      return scope.get(MetaCollectionImpl);
    });

    bind(MetaFactory)
      .toSelf()
      .inSingletonScope()
      .onActivation((context, factory) =>
        factory.register(TYPE_META_COMMENT, (model, position) => new MetaCommentImpl(model, position)),
      );

    bind(MenuItemFactory).toFactory(({ container }) => (model: MenuItemModel) => {
      const scope = container.createChild();
      scope.bind(MenuItemImpl).toSelf();
      scope.bind(MenuItemModelToken).toConstantValue(model);

      return scope.get(MenuItemImpl);
    });

    bind(ImageFactory).toFactory(({ container }) => (model: ImageModel) => {
      const scope = container.createChild();
      scope.bind(ImageImpl).toSelf();
      scope.bind(ImageModelToken).toConstantValue(model);

      return scope.get(ImageImpl);
    });

    bind(PaginationItemFactory).toFactory(({ container }) => (model: PaginationItemModel) => {
      const scope = container.createChild();
      scope.bind(PaginationItemImpl).toSelf();
      scope.bind(PaginationItemModelToken).toConstantValue(model);

      return scope.get(PaginationItemImpl);
    });

    bind(ContentFactory)
      .toSelf()
      .inSingletonScope()
      .onActivation(({ container }, factory) =>
        factory
          .register(TYPE_DOCUMENT, (model) => {
            const scope = container.createChild();
            scope.bind(DocumentImpl).toSelf();
            scope.bind(DocumentModelToken).toConstantValue(model);

            return scope.get(DocumentImpl);
          })
          .register(TYPE_IMAGE_SET, (model) => {
            const scope = container.createChild();
            scope.bind(ImageSetImpl).toSelf();
            scope.bind(ImageSetModelToken).toConstantValue(model);

            return scope.get(ImageSetImpl);
          })
          .register(TYPE_MENU, (model) => {
            const scope = container.createChild();
            scope.bind(MenuImpl).toSelf();
            scope.bind(MenuModelToken).toConstantValue(model);

            return scope.get(MenuImpl);
          })
          .register(TYPE_PAGINATION, (model) => {
            const scope = container.createChild();
            scope.bind(PaginationImpl).toSelf();
            scope.bind(PaginationModelToken).toConstantValue(model);

            return scope.get(PaginationImpl);
          }),
      );

    bind(ComponentFactory)
      .toSelf()
      .inSingletonScope()
      .onActivation(({ container }, factory) =>
        factory
          .register(TYPE_COMPONENT, (model, children) => {
            const scope = container.createChild();
            scope.bind(ComponentImpl).toSelf();
            scope.bind(ComponentModelToken).toConstantValue(model);
            scope.bind(ComponentChildrenToken).toConstantValue(children);

            return scope.get(ComponentImpl);
          })
          .register(TYPE_COMPONENT_CONTAINER, (model, children) => {
            const scope = container.createChild();
            scope.bind(ContainerImpl).toSelf();
            scope.bind(ComponentModelToken).toConstantValue(model);
            scope.bind(ComponentChildrenToken).toConstantValue(children);

            return scope.get(ContainerImpl);
          })
          .register(TYPE_COMPONENT_CONTAINER_ITEM, (model) => {
            const scope = container.createChild();
            scope.bind(ContainerItemImpl).toSelf();
            scope.bind(ComponentModelToken).toConstantValue(model);

            return scope.get(ContainerItemImpl);
          }),
      );

    bind(PageFactory).toFactory(({ container }) => (model: PageModel) => {
      const scope = container.createChild();
      scope.bind(PageImpl).toSelf();
      scope.bind(PageModelToken).toConstantValue(model);

      return scope.get(PageImpl);
    });
  });
}
