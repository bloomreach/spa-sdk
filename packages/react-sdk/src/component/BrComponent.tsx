/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isComponent, Component } from '@bloomreach/spa-sdk';
import { BrComponentContext } from './BrComponentContext';
import { BrNode } from './BrNode';

interface BrComponentProps {
  path?: string;
}

interface BrComponentState {
  components: Component[];
}

export class BrComponent extends React.Component<BrComponentProps, BrComponentState> {
  static contextType = BrComponentContext;

  constructor(props: BrComponentProps, public context: React.ContextType<typeof BrComponentContext>) {
    super(props, context);

    this.state = {
      components: this.getComponents(),
    };
  }

  componentDidUpdate(prevProps: BrComponentProps) {
    if (this.props.path !== prevProps.path) {
      this.setState({ components: this.getComponents() });
    }
  }

  private getComponents() {
    if (!this.context) {
      return [];
    }
    if (!this.props.path) {
      return this.context.getChildren();
    }

    const component = this.context.getComponent(...this.props.path.split('/'));

    return component ? [component] : [];
  }

  private renderComponents() {
    return this.state.components.map((component, index) => (
      <BrNode key={index} component={component}>{this.props.children}</BrNode>
    ));
  }

  render() {
    return <>{this.renderComponents()}</>;
  }
}
