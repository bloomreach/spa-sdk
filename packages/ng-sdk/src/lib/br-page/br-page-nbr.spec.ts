/*
 * Copyright 2020-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Component, NO_ERRORS_SCHEMA, OnInit } from '@angular/core';
import { ComponentFixture, fakeAsync, flush, TestBed } from '@angular/core/testing';
import { Component as BrComponent, initialize, Page } from '@bloomreach/spa-sdk';
import { BrSdkModule } from '../br-sdk.module';
import { BrPageComponent } from './br-page.component';

jest.mock('@bloomreach/spa-sdk');

describe('BrPageComponent', () => {
  let fixture: ComponentFixture<WrapperComponent>;
  let component: WrapperComponent;
  let componentEl: HTMLElement;

  const someEffect = jest.fn();

  const brComponent: BrComponent = {
    getId: jest.fn(),
    getMeta: jest.fn(),
    getModels: jest.fn(),
    getUrl: jest.fn(),
    getName: jest.fn(),
    getParameters: jest.fn(),
    getProperties: jest.fn(),
    getChildren: jest.fn(),
    getComponent: jest.fn(),
    getComponentById: jest.fn(),
  };
  const page: Page = {
    getButton: jest.fn(),
    getChannelParameters: jest.fn(),
    getComponent: jest.fn(() => brComponent),
    getContent: jest.fn(),
    getDocument: jest.fn(),
    getLocale: jest.fn(),
    getMeta: jest.fn(),
    getTitle: jest.fn(),
    getUrl: jest.fn(),
    getVersion: jest.fn(),
    getVisitor: jest.fn(),
    getVisit: jest.fn(),
    isPreview: jest.fn(),
    rewriteLinks: jest.fn(),
    sync: jest.fn(),
    toJSON: jest.fn(),
    sanitize: jest.fn(),
    prepareHTML: jest.fn(),
  };

  @Component({
    selector: 'br-my-component',
    template: '<div>Hello World</div>',
  })
  class MyComponent implements OnInit {
    ngOnInit(): void {
      someEffect();
    }
  }

  @Component({
    selector: 'br-wrapper-component',
    template: `
      <div>
        <br-page [configuration]="configuration" [mapping]="mapping">
          <ng-template>
            <br-my-component></br-my-component>
          </ng-template>
        </br-page>
      </div>
    `,
  })
  class WrapperComponent {
    configuration: BrPageComponent['configuration'] = {
      NBRMode: false,
    };

    mapping = {};
  }

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [MyComponent, WrapperComponent],
      imports: [HttpClientTestingModule, BrSdkModule],
      schemas: [NO_ERRORS_SCHEMA],
    });

    jest.mocked(initialize).mockImplementation(
      () => new Promise((resolve) => {
        setTimeout(() => {
          resolve(page);
        }, 1000);
      }),
    );

    fixture = TestBed.createComponent(WrapperComponent);
    component = fixture.componentInstance;
    componentEl = fixture.nativeElement;
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('template render', () => {
    it('should not render children if there is no page and NBR mode is false', fakeAsync(() => {
      fixture.detectChanges();
      expect(someEffect).not.toHaveBeenCalled();
      expect(componentEl).toMatchSnapshot();
      flush();
    }));

    it('should render children if there is no page and NBR mode is true', fakeAsync(() => {
      component.configuration = {
        NBRMode: true,
      };

      fixture.detectChanges();
      expect(someEffect).toHaveBeenCalled();
      expect(componentEl).toMatchSnapshot();
      flush();
    }));

    it('should render children if NBR mode is false and the PageModel becomes available', fakeAsync(() => {
      fixture.detectChanges();
      flush();
      fixture.detectChanges();
      expect(someEffect).toHaveBeenCalled();
    }));

    it('should initialize children only once if NBR mode is true and the PageModel becomes available', fakeAsync(() => {
      fixture.detectChanges();
      flush();
      fixture.detectChanges();
      expect(someEffect).toHaveBeenCalledTimes(1);
    }));
  });
});
