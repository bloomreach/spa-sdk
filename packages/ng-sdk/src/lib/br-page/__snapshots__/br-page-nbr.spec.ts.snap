// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrPageComponent template render should not render children if there is no page and NBR mode is false 1`] = `
<div
  id="root0"
>
  <div>
    <br-page
      ngskiphydration="true"
    >
      
      
    </br-page>
  </div>
</div>
`;

exports[`BrPageComponent template render should render children if there is no page and NBR mode is true 1`] = `
<div
  id="root1"
>
  <div>
    <br-page
      ngskiphydration="true"
    >
      
      
      
      <br-my-component>
        <div>
          Hello World
        </div>
      </br-my-component>
      
      
      
      
      
      
    </br-page>
  </div>
</div>
`;
